presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -223,11 +224,22 @@ public PlanWithProperties visitOutput(OutputNode node, Context context)
         {
             PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.undistributed()));
 
+            // if the child is already a gathering exchange, don't add another exchange
+            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType() == ExchangeNode.Type.GATHER) {
```

Comment:
> GATHER exchange does not necessarily mean `child.getProperties().isCoordinatorOnly()==true`

The reason this check is at this point is because even for the case of `if (!child.getProperties().isSingleNode())`, we don't want to  add another gathering exchange if child itself is a gathering exchange.

> Also, you are checking child of a child (grandchild) and not the child itself.

why do you say so? `child` is the node returned by `planChild()`, which indeed is the child. Isn't it?